# Usage of tzolkin-calendar in Your Python Code

For more information, visit the [Github project page](https://github.com/Release-Candidate/tzolkin-calendar)


## Import the Module

To use the Tzolk’in date package, import `tzolkin_calender` (with an underscore `_`).

A short check to see if it is working, is to print the version of `tzolkin-calendar`, that's the constant `tzolkin_calendar.VERSION`.

In [1]:
# Import the package.
import tzolkin_calendar

# Check, if it is working.
tzolkin_calendar.VERSION

'0.9.6'

## The Tzolk’in Date Class `Tzolkin`

The Tzolk’in date class resides in the module `tzolkin_calendar.tzolkin`, and is named `Tzolkin`.

So, import that module:

In [2]:
# Import the module tzolkin that contains `Tzolkin`.
from tzolkin_calendar import tzolkin

### Convert Gregorian Dates to Tzolk’in Dates

To get the Tzolk’in date of today, call the static method `fromToday`. This returns a `Tzolkin` instance holding the 
Tzolk’in date of today ('today' is the 22nd of March, 2021, with a Tzolk’in date of '1 Bʼen')

In [3]:
tzolkin.Tzolkin.fromToday()

4 Kʼibʼ

You can generate a `Tzolkin` instance from any gregorian date using the `datetime.date` class or a date string.

So, first import the datetime module:

In [4]:
import datetime

And then use the 3 possibilities to set the Tzolk’in date from a gregorian date. Or, in other words, to convert a gregorian date to a Tzolk’in date.

1. from a datetime.date instance using `Tzolkin.fromDate`. We use the method `fromisoformat` to set the gregorian date to the
22nd of March 2021 with the date string '2021-03-22'. The Tzolk’in date of the 22nd of March 2021 is '1 Bʼen'.

In [5]:
gregorian_date = datetime.date.fromisoformat("2021-03-22")

tzolkin.Tzolkin.fromDate(gregorian_date)

1 Bʼen

2. from an ISO date string using `Tzolkin.fromIsoFormat`. We set the Tzolk’in date to the 22nd of March 2021 with the ISO date
string '2021-03-22'. The Tzolk’in date of the 22nd of March 2021 is '1 Bʼen'.

In [6]:
tzolkin.Tzolkin.fromIsoFormat("2021-03-22")

1 Bʼen

3. from an arbitrary date string using `Tzolkin.fromDateString`. We set the Tzolk’in date to the 22nd of March 2021 with the date string '22=03\*2021' and the format string `fmt` '%d=%m*%Y'. The Tzolk’in date of the 22nd of March 2021 is '1 Bʼen'.

In [7]:
tzolkin.Tzolkin.fromDateString("22=03*2021", fmt="%d=%m*%Y")

1 Bʼen

### Set Tzolk’in Dates

You can set the `Tzolkin` instance to a Tzolk’in Date using it's constructor. The construcotr takes the Tzolk’in day number (between 1 and 13 including 1 and 13) and either a Tzolk’in day name or the number of the Tzolk’in day name (between 1 and 20 , including 1 and 20).

To get a dictionary of Tzolk’in day names and numbers, look at `tzolkin.day_names`.

In [8]:
tzolkin.day_names

{1: 'Imix',
 2: 'Ikʼ',
 3: 'Akʼbʼal',
 4: 'Kʼan',
 5: 'Chikchan',
 6: 'Kimi',
 7: 'Manikʼ',
 8: 'Lamat',
 9: 'Muluk',
 10: 'Ok',
 11: 'Chuwen',
 12: 'Ebʼ',
 13: 'Bʼen',
 14: 'Ix',
 15: 'Men',
 16: 'Kʼibʼ',
 17: 'Kabʼan',
 18: 'Etzʼnabʼ',
 19: 'Kawak',
 20: 'Ajaw'}

If we want to set a Tzolk’in day of '8 Kabʼan', we can either pass the day number 8 and day name Kabʼan to the constructor, or
the day number 8 and the day name number 17.

In [9]:
tzolkin.Tzolkin(number=8, name_str="Kabʼan")

8 Kabʼan

In [10]:
tzolkin.Tzolkin(number=8, name_number=17)

8 Kabʼan

If we pass an invalid number (not in [1, 13]) or name to the constructor, we get a `TzolkinException`.

```python
tzolkin.Tzolkin(number=53, name_number=17)

    TzolkinException: number 53 is not a valid Tzolkin day number, not between 1 and 13 (including 1 and 13)

tzolkin.Tzolkin(number=3, name_str="Hugo")

    TzolkinException: string "Hugo" is not a valid Tzolkin day name, one of: dict_values(['Imix', 'Ikʼ', 'Akʼbʼal', 'Kʼan', 'Chikchan', 'Kimi', 'Manikʼ', 'Lamat', 'Muluk', 'Ok', 'Chuwen', 'Ebʼ', 'Bʼen', 'Ix', 'Men', 'Kʼibʼ', 'Kabʼan', 'Etzʼnabʼ', 'Kawak', 'Ajaw'])
          
tzolkin.Tzolkin(number=3, name_number=-5)

    TzolkinException: -5 is not a valid Tzolkin day name number, it must be between 1 and 20 (including 1 and 20)
```

These Tzolk’in day numbers and names can be accessed using the mthods `getDayNumber`, `getDayName` and `getDayNameNumber`.

In [11]:
# Set the Tzolk’in date to '12 Kimi'.
tzolkin_date = tzolkin.Tzolkin(number=12, name_str="Kimi")

tzolkin_date.getDayNumber()

12

In [12]:
tzolkin_date.getDayName()

'Kimi'

In [13]:
tzolkin_date.getDayNameNumber()

6

 To get the number of Tzolk’in day in the Tzolk’in year of 260 days, there is `getTzolkinYearDay`. For example '12 Kimi' is the 246. day (of 260 days)  of the Tzolk’in year

In [14]:
tzolkin_date.getTzolkinYearDay()

246

To 

In [15]:
day_number = tzolkin.Tzolkin.parseTzolkinName("EtZ`nAB")
if day_number != 0:
    tzolkin_name = tzolkin_calendar.day_names[day_number]
tzolkin_name

'Etzʼnabʼ'

All 260 Tzolk’in days of a Tzolk’in year we can get as a list of strings from the static method `getTzolkinCalendar`.

In [16]:
tzolkin.Tzolkin.getTzolkinCalendar()

['1 Imix',
 '2 Ikʼ',
 '3 Akʼbʼal',
 '4 Kʼan',
 '5 Chikchan',
 '6 Kimi',
 '7 Manikʼ',
 '8 Lamat',
 '9 Muluk',
 '10 Ok',
 '11 Chuwen',
 '12 Ebʼ',
 '13 Bʼen',
 '1 Ix',
 '2 Men',
 '3 Kʼibʼ',
 '4 Kabʼan',
 '5 Etzʼnabʼ',
 '6 Kawak',
 '7 Ajaw',
 '8 Imix',
 '9 Ikʼ',
 '10 Akʼbʼal',
 '11 Kʼan',
 '12 Chikchan',
 '13 Kimi',
 '1 Manikʼ',
 '2 Lamat',
 '3 Muluk',
 '4 Ok',
 '5 Chuwen',
 '6 Ebʼ',
 '7 Bʼen',
 '8 Ix',
 '9 Men',
 '10 Kʼibʼ',
 '11 Kabʼan',
 '12 Etzʼnabʼ',
 '13 Kawak',
 '1 Ajaw',
 '2 Imix',
 '3 Ikʼ',
 '4 Akʼbʼal',
 '5 Kʼan',
 '6 Chikchan',
 '7 Kimi',
 '8 Manikʼ',
 '9 Lamat',
 '10 Muluk',
 '11 Ok',
 '12 Chuwen',
 '13 Ebʼ',
 '1 Bʼen',
 '2 Ix',
 '3 Men',
 '4 Kʼibʼ',
 '5 Kabʼan',
 '6 Etzʼnabʼ',
 '7 Kawak',
 '8 Ajaw',
 '9 Imix',
 '10 Ikʼ',
 '11 Akʼbʼal',
 '12 Kʼan',
 '13 Chikchan',
 '1 Kimi',
 '2 Manikʼ',
 '3 Lamat',
 '4 Muluk',
 '5 Ok',
 '6 Chuwen',
 '7 Ebʼ',
 '8 Bʼen',
 '9 Ix',
 '10 Men',
 '11 Kʼibʼ',
 '12 Kabʼan',
 '13 Etzʼnabʼ',
 '1 Kawak',
 '2 Ajaw',
 '3 Imix',
 '4 Ikʼ',
 '5 Akʼbʼal',
 '6 K

### Search Gregorian Dates to a given Tzolk’in Date

We can search for the next (forward in time) or last (backwards in time) day with the same Tzolk’in date using the methods `getNextDate` and `getLastDate`. Both methods return a `datetime.date` object.

When searching for the next gregorian date that has the Tzolk’in date '7 Kawak', we get the 28th of March, 2021 - because we started searching 'today', which is the 22nd of March 2021.

In [17]:
# Set the Tzolk’in date to search for to '7 Kawak'.
tzolkin_date = tzolkin.Tzolkin(number=7, name_str="Kawak")

tzolkin_date.getNextDate()

datetime.date(2021, 3, 28)

When searching for the last gregorian date that has the Tzolk’in date '7 Kawak', we get the 11th of July, 2020 - because we started searching 'today', which is the 22nd of March 2021.

In [18]:
tzolkin_date.getLastDate()

datetime.date(2020, 7, 11)

Both methods, `getNextDate` and `getLastDate` take an optional argument `start_date`, which is the gregorian date to start
the search. If no `start_date` is given, 'today' is used as the start date.

So now we search again for '7 Kawak' in both directions, but this time we start at the 10th of July, 2020.

In [19]:
# Set the Tzolk’in date to search for to '7 Kawak'.
tzolkin_date = tzolkin.Tzolkin(number=7, name_str="Kawak")

# Set the start date of the search to the 10th of July, 2020.
start_search = datetime.date.fromisoformat("2020-07-10")

start_search.isoformat()

'2020-07-10'

For the next day with a Tzolk’in date of '7 Kawak' we now get the 11th of July, 2020.

In [20]:
tzolkin_date.getNextDate(start_date=start_search)

datetime.date(2020, 7, 11)

For the last day before our start date with a Tzolk’in date of '7 Kawak' we now get the 25th of October, 2019.

In [21]:
tzolkin_date.getLastDate(start_date=start_search)

datetime.date(2019, 10, 25)

To get a list of `datetime.date` dates with the same Tzolk’in date, we can use the methods `getNextDateList` and `getLastDateList`.

Again, we can set the argument `start_date` to a gregorian date to start the search or not set it to start the search today.
The number of elements in the returned list is set using the parameter `list_size`, which defaults to 50.

In [22]:
# Set the Tzolk’in date to search for to '7 Kawak'.
tzolkin_date = tzolkin.Tzolkin(number=7, name_str="Kawak")

Let's start the search for dates with a Tzolk’in date of ' 7 Kawak' today, the 22nd of March 2021, and set the list size to 9 elements:

In [23]:
tzolkin_date.getNextDateList(list_size=9)

[datetime.date(2021, 3, 28),
 datetime.date(2021, 12, 13),
 datetime.date(2022, 8, 30),
 datetime.date(2023, 5, 17),
 datetime.date(2024, 2, 1),
 datetime.date(2024, 10, 18),
 datetime.date(2025, 7, 5),
 datetime.date(2026, 3, 22),
 datetime.date(2026, 12, 7)]

Now start searching for '7 Kawak' on the 29th of March, 2021 and set the returned list size to 5.

In [24]:
# Set the start date of the search to the 29th of March, 2021.
start_search = datetime.date.fromisoformat("2021-03-29")

tzolkin_date.getLastDateList(start_date=start_search, list_size=5)

[datetime.date(2021, 3, 28),
 datetime.date(2020, 7, 11),
 datetime.date(2019, 10, 25),
 datetime.date(2019, 2, 7),
 datetime.date(2018, 5, 23)]

### Calculations using  Tzolk’in Dates

There are 4 methods to get the difference in days between two Tzolk’in dates and to add (or subtract) days from a Tzolk’in date: `addDays`, `addTimedelta`, `getDayDiff` and `getDayTimedelta`.

Lets start with a Tzolk’in date of '6 Muluk'.

In [25]:
tzolkin.Tzolkin(number=6, name_str="Muluk")

6 Muluk

Add 6 days to it, and we get a Tzolk’in date of '12 Men'.

In [26]:
tzolkin.Tzolkin(number=6, name_str="Muluk").addDays(6)

12 Men

Instead of using ints, we can also add and subtract `datetime.timedelta` objects.
Now subtract 6 days from '12 Men' - we get '6 Muluk'.

In [27]:
to_subtract = datetime.timedelta(days=-6)

tzolkin.Tzolkin(number=12, name_str="Men").addTimedelta(to_subtract)

6 Muluk

To get the difference between two Tzolk’in dates there exist the Methods `getDayDiff` and `getDayTimedelta`.

Lets calculate the difference in days between '6 Muluk' and '12 Men'.

In [28]:
# Set start_tzolkin to '6 Muluk'.
start_tzolkin = tzolkin.Tzolkin(number=6, name_str="Muluk")

# Set end_tzolkin to '12 Men'.
end_tzolkin = tzolkin.Tzolkin(number=12, name_str="Men")

start_tzolkin.getDayDiff(end_tzolkin)

6

And using `getDayTimedelta`, which returnes a `datetime.timedelta` object.

In [29]:
start_tzolkin.getDayTimedelta(end_tzolkin)

datetime.timedelta(days=6)

What happens, if we calculate the difference between '12 Men' and '6 Muluk'?

In [30]:
end_tzolkin.getDayDiff(start_tzolkin)

254

We get 254 days, not -6. That's because the difference is always calculated forward in time. If you want to get negative days or the shortest possible time difference, subtract 260 from the the result (the number of days in a Tzolk’in year). As soon as the difference in days is greater than 130, to minimum time distance in days 'is negative'.

In [31]:
day_diff = end_tzolkin.getDayDiff(start_tzolkin)

if day_diff > 130:
    day_diff = 260 - day_diff
day_diff

6

Or you can use the minimum of `result` and ``|result - 260|`` that is ``abs(result - 260)``.

In [32]:
day_diff = end_tzolkin.getDayDiff(start_tzolkin)
shortest_diff = min(day_diff, abs(day_diff - 260))
shortest_diff

6

This Notebook is licensed under the MIT License, see [LICENSE](https://github.com/Release-Candidate/tzolkin-calendar/blob/main/LICENSE)
&copy; Roland Csaszar 2021